lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -2303,14 +2303,25 @@ SILValue LifetimeChecker::handleConditionalInitAssign() {
       B.setInsertionPoint(TrueBB->begin());
       SILValue EltPtr;
       {
-        llvm::SmallVector<std::pair<SILValue, SILValue>, 4> EndBorrowList;
-        EltPtr = TheMemory.emitElementAddress(Elt, Loc, B, EndBorrowList);
+        using EndScopeKind = DIMemoryObjectInfo::EndScopeKind;
+        SmallVector<std::pair<SILValue, EndScopeKind>, 4> EndScopeList;
+        EltPtr =
+            TheMemory.emitElementAddressForDestroy(Elt, Loc, B, EndScopeList);
         if (auto *DA = B.emitDestroyAddrAndFold(Loc, EltPtr))
           Destroys.push_back(DA);
-        while (!EndBorrowList.empty()) {
-          SILValue Borrowed, Original;
-          std::tie(Borrowed, Original) = EndBorrowList.pop_back_val();
-          B.createEndBorrow(Loc, Borrowed, Original);
+        for (auto p : EndScopeList) {
```

Comment:
Looking at this again, I need to think about if I need to pop off from the back instead to get the ordering right.

Reply:
Yea, I think I am wrong to do it this way. I think the mistake I saw before was that I was not inserting the cleanup for a borrow into the stack at the appropriate time. For the present it shouldn't be an issue, but I am going to fix before merging.